### PR TITLE
fix erase_if method

### LINKED ISSUE #17043: [Regression] Infamous 2
### Quick summary

Infamous 2 crashes randomly

### Details

Infamous 2 keeps crashing after some play time, why precisely i can't tell, time varies but it eventually happens: sometimes a few minutes, sometimes a single minute

### Build with regression

v0.0.36-17781-7500d165

### Attach two log files

[17781 bad.log](https://github.com/user-attachments/files/19730243/17781.bad.log)

[17779 good.log](https://github.com/user-attachments/files/19730304/17779.good.log)

### Attach capture files for visual issues

_No response_

### System configuration

Windows 10 22H2 19045.5737
Intel i3-10105
Nvidia RTX 3060 12GB
Nvidia driver version 572.83
24 GB DDR4
RPCS 3 running on 1tb HDD, Infamous 2 files on 1tb SSD

### Other details

thank you

### LINKED ISSUE #17098: [Regression] Resident Evil Operation Raccoon City
### Quick summary

The build with the regression crashed on the press start screen.

### Details

The build regressed the game from in game status to intro since it crashes on the very moment the press start screen appears right after showing the logos for the companies that made it.

### Build with regression

RPCS3 v0.0.36-17781-7500d165 

### Attach two log files

[RPCS3 17779 Working.log.gz](https://github.com/user-attachments/files/19914881/RPCS3.17779.log.gz)

[RPCS3 17781 Broken.log](https://github.com/user-attachments/files/19914878/RPCS3.broken.RE.ORC.log)

### Attach capture files for visual issues

_No response_

### System configuration

Intel Core i5-14600K | 20 Threads | 31.78 GiB RAM | AVX+ | FMA3
GPU: NVIDIA GeForce RTX 3070 (576.02)
Windows 11 Professional

### Other details

It crashed without showing any errors. The emulator just kept running in the background eating up resources even after I closed the application window I had to force close it through Task Manager.

### PR DESCRIPTION
**Pull request by digant73**
Fixes made by #17040 highlighted a further bug in `erase_if` method that should be fixed now. Probably more than the listed bug reports below were affected by the bug.

fixes #17098
should fix #17043

### CODE REVIEW

File: rpcs3/Emu/RSX/Common/simple_array.hpp [hunk 1]
```diff
@@ -405,7 +405,7 @@ namespace rsx
 
 					// Move item to the end of the list and shrink by 1
 					std::memcpy(ptr, last, sizeof(Ty));
-					last = _data + (--_size);
+					last = _data + (--_size - 1); // set new last
```
[the remark this answers is not in the document]
Comment (digant73) [reply rpcs3/Emu/RSX/Common/simple_array.hpp:408]: yes I also thought that the logic was overall not correct. The swapped ptr <> last is not tested (ptr will be increased in the `for` cycle

[the remark this answers is not in the document]
Comment (digant73) [reply rpcs3/Emu/RSX/Common/simple_array.hpp:408]: if scanning/swapping order is not important then the following code should be ok:

```
			for (auto last = _data + (_size - 1), ptr = last; ptr >= _data; ptr--)
			{
				if (predicate(*ptr))
				{
					ret = true;

					if (ptr != last) // if not last
					{
						// Move item to the end of the list
						std::memcpy(ptr, last, sizeof(Ty));
					}

					// Popping the last entry from list and set new last
					_size--;
					last--;
				}
			}

```

[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/Common/simple_array.hpp:408]: https://github.com/RPCS3/rpcs3/pull/17102 fixes this properly. I also wrote the test for it in case you want to toy around with it: https://godbolt.org/z/4745MGT6z

[the remark this answers is not in the document]
Comment (digant73) [reply rpcs3/Emu/RSX/Common/simple_array.hpp:408]: not sure I understood the missing removal, it seems there, but ok I see you provided the PR, so closing this

[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/Common/simple_array.hpp:408]: A better test would include a larger random list with repeating elements and such but I'm under some time pressure at the moment. The solution should be robust enough for now to handle the edge cases correctly.

[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/Common/simple_array.hpp:408]: > not sure I understood the missing removal, it seems there, but ok I see you provided the PR, so closing this

My fault for not testing this thoroughly, I missed the reversed direction of the scan.

[the remark this answers is not in the document]
Comment (digant73) [reply rpcs3/Emu/RSX/Common/simple_array.hpp]: pushed my changes as a reminder in case they will be ok.
I also thought if the items removed from the list are object types needing to be destroyed (e.g. release allocated pointers etc.)

[the remark this answers is not in the document]
Comment (kd-11) [reply rpcs3/Emu/RSX/Common/simple_array.hpp]: Simple array is stricly for POD types. This is enforced by the template.
Now for the impl, yours is fine so I'll merge to give credit, but I absolutely despise reversed loops unless I have no other choice (debugging requires inverting logic in the head which is not easy) so I will also merge my PR after to replace it.

[the remark this answers is not in the document]
Comment (digant73) [reply rpcs3/Emu/RSX/Common/simple_array.hpp]: no problem at all. Not interested to credits etc.. Any solution fixing bugs is ok for me. I simply pushed my last changes just as reminder for myself before (sooner or later) I will permanently delete this PR from my repo.
